calyx/src/passes/group_to_invoke.rs

```diff
@@ -31,100 +34,190 @@ impl Named for GroupToInvoke {
+                inputs.push((name, Rc::clone(&assign.src)));
+            } else {
+                // assign has a guard condition,so need a wire
+                let width = assign.dst.borrow().width;
+                let wire =
+                    builder.add_primitive("std_wire", "std_wire", &[width]);
+                let wire_in_rrc = wire.borrow().get("in");
+                let asmt = builder.build_assignment(
+                    wire_in_rrc,
+                    Rc::clone(&assign.src),
+                    *assign.guard.clone(),
+                );
+                comb_assigns.push(asmt);
+                let wire_out_rrc = wire.borrow().get("out");
+                inputs.push((name, wire_out_rrc));
+            }
         }
     }
 
-    ir::Control::invoke(comp, inputs, outputs)
+    let comb_group = if comb_assigns.is_empty() {
+        None
+    } else {
+        let comb_group_ref = builder.add_comb_group("comb_invoke");
+        comb_group_ref
+            .borrow_mut()
+            .assignments
+            .append(&mut comb_assigns);
+        Some(comb_group_ref)
+    };
+
+    ir::Control::Invoke(ir::Invoke {
+        comp,
+        inputs,
+        outputs: Vec::new(),
+        comb_group,
+        attributes: ir::Attributes::default(),
+        ref_cells: Vec::new(),
+    })
 }
 
 impl Visitor for GroupToInvoke {
-    fn enable(
+    fn start(
         &mut self,
-        s: &mut ir::Enable,
-        _comp: &mut ir::Component,
-        _sigs: &ir::LibrarySignatures,
+        comp: &mut ir::Component,
+        sigs: &ir::LibrarySignatures,
         _comps: &[ir::Component],
     ) -> VisResult {
-        let group = s.group.borrow();
-
-        // There should be exactly one component being written to in the
-        // group.
-        let mut writes =
-            ReadWriteSet::write_set(group.assignments.iter()).collect_vec();
-        if writes.len() != 1 {
-            return Ok(Action::Continue);
-        }
-
-        // Component must define a @go/@done interface
-        let cell = writes.pop().unwrap();
-        let maybe_go_port = cell.borrow().find_with_attr("go");
-        let maybe_done_port = cell.borrow().find_with_attr("done");
-        if maybe_go_port.is_none() || maybe_done_port.is_none() {
-            return Ok(Action::Continue);
-        }
+        let groups: Vec<ir::RRC<ir::Group>> = comp.groups.drain().collect();
+        let mut builder = ir::Builder::new(comp, sigs);
+        for g in groups.iter() {
+            let group = g.borrow();
+            let mut writes: Vec<ir::RRC<ir::Cell>> =
+                ReadWriteSet::write_set(group.assignments.iter())
+                    .filter(|cell| match cell.borrow().prototype {
+                        ir::CellType::Primitive { is_comb, .. } => !is_comb,
+                        _ => true,
+                    })
+                    .collect_vec();
+            // should only have one write to a non-combinational component
```

Comment:
Why is this required? A `comb group` can have guarded assignments to a cell in it

Reply:
I meant "should only be 1 write, among writes to non-combinational components". I'll fix it to say "Excluding writes to combinational components, should only write to one cell".